cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -131,23 +130,163 @@ __device__ __host__ double attractive_grad(double dist_squared,
  * randomly.
  */
 template <typename T, int TPB_X>
-__global__ void optimize_batch_kernel(
+__global__ void optimize_batch_kernel_on_shared_mem(
   T *head_embedding, int head_n, T *tail_embedding, int tail_n, const int *head,
   const int *tail, int nnz, T *epochs_per_sample, int n_vertices,
   bool move_other, T *epochs_per_negative_sample,
   T *epoch_of_next_negative_sample, T *epoch_of_next_sample, double alpha,
   int epoch, double gamma, uint64_t seed, UMAPParams params) {
+  extern __shared__ T embedding_updates[];
+
+  int row = (blockIdx.x * TPB_X) + threadIdx.x;
+  if (row < nnz) {
+    /**
+     * Positive sample stage (attractive forces)
+     */
+
+    if (epoch_of_next_sample[row] <= epoch) {
+      int j = head[row];
+      int k = tail[row];
+
+      T *current = head_embedding + (j * params.n_components);
+      T *other = tail_embedding + (k * params.n_components);
+      T *current_buffer, *other_buffer;
+
+      if (move_other) {
+        current_buffer =
+          embedding_updates + ((threadIdx.x * 2) * params.n_components);
```

Comment:
Your question this morning about the `__syncthreads` makes total sense now. If each thread is just updating their own completely independent section of shared memory, I'm not sure a `__syncthreads` is necessary at all. In that case, I don't believe you would even need to do an atomic update to the shared memory, since no other threads will be writing to it.

One of the reasons I was thinking the shared memory approach would be so effective at boosting performance is because the threads would be writing to the same set of shared memory indices, based on the vertices rather than the thread ids. In that case, you would probably use thread0 to create an associative map of the unique vertices in he block to shared memory indices. I'm okay saving this particular optimization for the future. If we do that, we should probably remove the sync's and atomics.

Reply:
Yes, absolutely. I think I was somewhat mixing the two ideas in my mind, but did not implement the mapping. Totally agree that the first `__syncthreads()` and the atomicAdd on shared memory are in fact useless. The last one though prevents a thread from modifying the embedding before an other one would read it, at least locally to the block (not perfect as discussed). I think this method without mapping still avoids `(1 + n_neg_samples) * n_components` (+ an additional `n_components` for fit) accesses to global memory per thread, so it is still more efficient.

Reply:
> The last one though prevents a thread from modifying the embedding before an other one would read it, at least locally to the block

Agreed, the last `_syncthreads` should stay, along with the atomic adds to global memory. 

I also want to make sure we aren't assuming much with regards to the effects that using shared memory is intended to have on global consistency, since this is dependent on too many different factors (eg. the number of threads per block, number of data samples, number of neighbors, and the result of the weighting step). 

My intention for this shared memory trick, especially for its use in the consistent code path, is only to boost the performance enough to offset the effect of the double precision sums. This is primarily why I'd like to see it used in that code path (in addition to the synchronous updates).

Reply:
Yes, so I'll use shared memory for the consistent path too. Ok for the optimal method with mapping, we will open an issue to implement that later.